### ISSUE TITLE
[java] InvalidSlf4jMessageFormat false positive

### ISSUE THREAD

**Issue by mkordas** [pmd/pmd#935]
**Affects PMD Version:** 
6.0.1

**Rule:**
InvalidSlf4jMessageFormat

**Description:**
When exception is returned via method in the last argument it should not be raised as violation. 

**Code Sample demonstrating the issue:**
```java
	public static IllegalStateException exception() { return new IllegalStateException();}
	
	public static void log() {
                // violation raised below, but code is fine
		LOGGER.warn("Exception {}", IllegalStateException.class.getName(), exception());
	}
```

**Comment by adangel** [comment]
@mkordas Thanks for reporting. We currently can't determine the return type of the method call - we would need this, to ensure, the method actually returns an exception type and the argument doesn't count for the placeholders.

Without context, this code example looks very artificial - you usually get an exception in a catch clause, and then you have a local variable, with which the rule works already (because we now the types of the variables). Would you mind to share a bit more about the use case?

**Comment by jsotuyod** [comment]
@adangel we actually have the return type of the method call (provided it doesn't fall into the corner cases for type inference currently unsupported). We should be able to detect this scenario more often than not.

**Comment by boris-petrov** [comment]
I didn't know that you could give an exception as the last parameter to the overloads with a variable number of arguments. This is supposed to do the same as the overload that explicitly accepts an Exception as the last parameter, right?

I actually tried changing our codebase to use this style. However, I got a couple of false-positives from PMD. Whenever the last arguments that is passed is a `Throwable` (instead of an `Exception`), PMD gives a `Too many arguments, expected 1 argument but have 2` warning. Should I open a new issue for this?

**Comment by jsotuyod** [comment]
@boris-petrov yes, please report a separate issue for that. Thanks in advanced!

**Comment by mkordas** [comment]
@adangel Indeed, my example may look artificial. In the real code, violation was triggered in exception handling framework, where exceptions are stored as fields and retrieved via methods to log them. Code is too complex to post here, but I think now you understand the use case.

**Comment by zodac** [comment]
Commenting here as I'm seeing a similar issue, but only when replacing some code with a lambda.

**Affects PMD Version:**
6.10.0

**Running PMD through:**
Maven 3.6.0

Original code with anonymous class:

```
return new Callback() {
    @Override
    public void onCompletion(final RecordMetadata recordMetadata, final Exception e) {
        if (e == null) {
            LOGGER.debug("Record sent to '{}' with offset '{}'", recordMetadata.topic(), recordMetadata.offset());
        } else {
            LOGGER.error("Exception while trying to send a record '{}'", recordMetadata, e); // Does not cause a PMD violation
        }
    }
};
```

Code with lambda:

```
return (recordMetadata, e) -> {
    if (e == null) {
        LOGGER.debug("Record sent to '{}' with offset '{}'", recordMetadata.topic(), recordMetadata.offset());
    } else {
        LOGGER.error("Exception while trying to send a record '{}'", recordMetadata, e); // PMD violation for rule "InvalidSlf4jMessageFormat": Too many arguments, expected 1 argument but have 2.
    }
};
```

**Comment by jsotuyod** [comment]
All the cases cited on this issue are no longer reproducible in PMD 7.0.0.
